Extend extension API to support joystick suggestions and extra iFrame options

rafaellehmkuhl opened the pull request:
## Joystick Mapping Suggestions from Extensions

BlueOS extensions can now suggest joystick button mappings to the user.

When suggestions are available, a modal pops up showing them organized by extension and by group. Users can apply suggestions one by one or all at once, and can also ignore suggestions they don't want.

Each joystick profile tracks its own set of applied and ignored suggestions independently.

## Revamped External Features Discovery Modal

The modal that previously only handled action suggestions was expanded to also cover joystick mapping suggestions.

It was renamed from `ActionDiscoveryModal` to `ExternalFeaturesDiscoveryModal` and received a major UI overhaul:

- Suggestions are grouped by source extension with collapsible sections
- There's a joystick profile selector
- Bulk apply/ignore controls

## Safer BlueOS API Communication

Responses from BlueOS are now validated using schemas before being used, reducing the chance of crashes from unexpected data.

All external API types were also converted to the internal naming style for consistency.

## Cleaner Type Organization

Type definitions that were previously scattered across implementation files were moved to dedicated type files, making them easier to find, reuse, and maintain.

## Other Improvements

- **IFrame widgets** can now start in a collapsed state
- Support extra attributes from the extension API

Fix #1859

rafaellehmkuhl commented:
> Hi @rafaellehmkuhl, when I was implementing the API on my extension, I noticed that this implementation is using `camelCase` convention for the fields, which differs from the external widgets (see `ExternalWidgetSetupInfo`), and it is using `kebab-case` for the enum variants, which also differs from the external widgets (see `WidgetType` or `MiniWidgetType`).
> 
> It would be good if we could agree to stick to a convention for the entire cockpit_extras API.

Fixed the usage of `camelCase` in the API. It is all in `snake_case` now.

The enums will have to stay as they are for historical purposes, otherwise we would break already-stablished APIs.

joaoantoniocardoso commented:
> > Hi @rafaellehmkuhl, when I was implementing the API on my extension, I noticed that this implementation is using `camelCase` convention for the fields, which differs from the external widgets (see `ExternalWidgetSetupInfo`), and it is using `kebab-case` for the enum variants, which also differs from the external widgets (see `WidgetType` or `MiniWidgetType`).
> > It would be good if we could agree to stick to a convention for the entire cockpit_extras API.
> 
> Fixed the usage of `camelCase` in the API. It is all in `snake_case` now.
> 
> The enums will have to stay as they are for historical purposes, otherwise we would break already-stablished APIs.

Okay, but you could also make a parser that normalizes the old format to a new format, keeping backward compatibility while updating to this new uniform style.

rafaellehmkuhl committed:
packages: Add `zod`

rafaellehmkuhl committed:
packages: Add `camelcase-keys`

rafaellehmkuhl committed:
widgets: iframe: Allow it to start collapsed already

rafaellehmkuhl committed:
joystick: Concentrate info about joystick models in a single file

rafaellehmkuhl committed:
joystick: manager: Fix import sorting

rafaellehmkuhl committed:
data-lake: Segregate type definitions to a dedicated file

rafaellehmkuhl committed:
cockpit-actions: Remove duplicated types

rafaellehmkuhl committed:
cockpit-actions: Concentrate types on single file

rafaellehmkuhl committed:
cockpit-actions: Be more specefic about config type

rafaellehmkuhl committed:
blueos: external-widgets: Expose missing iFrame options

rafaellehmkuhl committed:
blueos: Refactor external resources importing

- Allow importing joystick mapping suggestions
- Segregate types to dedicated file
- Add schema validation
- Standardize payload keys to camelCase

ArturoManzoli reviewed (changes requested):
**Function and code-wise everything is working very well!**

Some UI tweaks are needed before merging:

(1) There is a large unused head space under the dialog's title. Reduce to 1/3 of that;
(2) The border is too white. Make it #ffffff14, the same as the glass element's border.
 <img width="860" height="563" alt="Image" src="https://github.com/user-attachments/assets/cf1c93d4-7adc-475f-833e-5355fc915dc1" />

(1) The blue tag is hard to see over some bright backgrounds, specially by colorblind and people with contrast difficulties. Change it to a darker blue filled pill with white text on it. 
<img width="919" height="545" alt="Image" src="https://github.com/user-attachments/assets/88b28514-e8a3-40ba-b1bf-da41e795f4fd" />

(1) There is unneeded vertical space taken here by the subtitle. You could remove it, since its almost redundant with the title, or, if its needed in other cases, add it to the same line the title is, but on a smaller font.
(2) Same vertical case here. Is it possible to make they all share the same row? We would free about 100 vertical pixels by doing so.
<img width="1141" height="1003" alt="Image" src="https://github.com/user-attachments/assets/96118255-a5e2-40ba-8131-9b1b83484275" />

(1) Spread (justify-between) the buttons on the lower part of the card. Elements will look even better.
<img width="389" height="459" alt="Image" src="https://github.com/user-attachments/assets/c9970736-c92e-447c-b4f6-121c268027d4" />

(1) On the following dialogs, both have its cancel button to the right, near the apply. Make them apart from each other.
<img width="542" height="583" alt="Image" src="https://github.com/user-attachments/assets/c0f21915-87b8-435c-a813-e205b84c3d6e" />
<img width="772" height="422" alt="Image" src="https://github.com/user-attachments/assets/51ad3a91-423a-444a-b0f2-f77ab1743612" />

rafaellehmkuhl committed:
cockpit-actions: Accept custom action IDs on creation

This allows for external agents to decide their own IDs and track that as they want.

rafaellehmkuhl commented:
> The API worked well, but I couldn't get the joystick to assign the L3 (one-push white balance), maybe because the action didn't exist?

Problem fixed.

rafaellehmkuhl committed:
yarn: Update lock file

ArturoManzoli commented on src/components/ExternalFeaturesDiscoveryModal.vue:
```diff
@@ -0,0 +1,1364 @@
+                            </template>
+                            <span class="text-xs">{{ group.description }}</span>
+                          </v-tooltip>
+                          <v-chip size="x-small" variant="tonal" color="blue-darken-2" text-color="white">
```
The native vuetify props color and text-color not always work (some others neither :)). 

Its safer to use on this case:
`<v-chip size="x-small" variant="tonal" class="bg-[#20538d] text-white font-bold">`

ArturoManzoli reviewed (changes requested):
Just a small change on the v-chip color, as suggested, and we're ready to go in my opinion

All the rest is fine!

ArturoManzoli reviewed (approved):
All good!
